libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c

```diff
@@ -141,6 +162,63 @@ TEST( Provisioning_Unit_Serializer, TestSerializeCreateKeysAndCertificatePayload
     AwsIotProvisioning_FreePayload( pSerializationBuffer );
 }
 
+/**
+ * @brief Tests that the CSR payload serializer calculates the serialization size, and populates the passed output
+ * parameter when a serialization buffer is not passed to it.
+ */
+TEST( Provisioning_Unit_Serializer, TestSerializeCreateCertFromCsrPayloadWithoutBuffer )
+{
+    uint8_t * pSerializationBuffer = NULL;
+    size_t bufferSizeNeeded = 0;
+
+    /* Test the serializer function. */
+    TEST_ASSERT_EQUAL( true,
+                       _AwsIotProvisioning_SerializeCreateCertFromCsrRequestPayload( _testCsrString,
+                                                                                     strlen( _testCsrString ),
+                                                                                     pSerializationBuffer,
+                                                                                     &bufferSizeNeeded ) );
+    /* Make sure that the output parameter has been populated with the serialization size. */
+    TEST_ASSERT_EQUAL( sizeof( _expectedSerialization ), bufferSizeNeeded );
+}
+
+/**
+ * @brief Tests that the CSR payload serializer populates the buffer with the serialized payload, when
+ * the passed buffer is valid.
+ */
+TEST( Provisioning_Unit_Serializer, TestSerializeCreateCertFromCsrPayloadWithBuffer )
+{
+    uint8_t testBuffer[ sizeof( _expectedSerialization ) ] = { 0 };
+    size_t bufferSize = sizeof( testBuffer );
+
+    /* Test the serializer function. */
+    TEST_ASSERT_EQUAL( true,
+                       _AwsIotProvisioning_SerializeCreateCertFromCsrRequestPayload( _testCsrString,
+                                                                                     strlen( _testCsrString ),
+                                                                                     &testBuffer[ 0 ],
```

Comment:
You can improve this type of test by reserving a few `uint8_t`s at the beginning and end of `testBuffer`, filling them with some known values (e.g. `0xa5`), and then calling your routine under test with `&testBuffer[ X ]`, where `X` is after your special filled area.  After calling your routine you can then verify that your special values are still in place at the beginning and end of the test array.  This can help catch some common overrun errors.

^ In general I like to capture that in a common test utility that is used throughout all code which does any sort of buffer formatting.

Reply:
That's a good suggestion. Incorporated it in the test